proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,280 @@
+slot for the BankHash computation.  This slot is different
+from the tower vote slot.
+
+Each validator that is computing a BankHash will be able to verify
+the BankHash values created by block producers for previous slots
+because each BankHash has a link to the previous one.
+
+During verification, if a validator sees that > 1/3 of the cluster
+has incorrectly computed a state transition, they should halt
+immediately.
+
+### VoteHash
+
+All validators will execute all the vote transactions on every fork,
+as they do now. After computing all the vote state transitions the
+validator will compute the VoteHash, identical to the BankHash in
+structure, but covers only the vote state transitions.  The VoteHash
+is included in the vote for the slot that the validator is voting
+for.
+
+### Client Confirmations
+
+Clients connected to a trusted RPC can confirm the transaction as
+soon as the RPC has executed teh UserBlock and the block has been
+optimistically confirmed.
+
+Client can also wait for a status code weighted by stake weight,
+as each validator's execution catches up to the slot and includes
+a BankHash confirming the state of the execution.
+
+It may be sufficient to wait for 1/3+ validators executing the state
+transition instead of the full 2/3+, because if 1/3+ are incorrect
+the cluster will halt anyways.
+
+### State synchronization
+
+Nodes must be able to compute a full snapshot at least once an
+epoch. So overall CU limits for the block must be set such that
+synchronous execution is possible. But because asynchronous execution
+is possible, replay can take advantage of much larger batches of
+execution.
+
+## Impact
+
+Multiple nodes can operate as Builder on the cluster concurrently.
+So clients can pick the nearest one, and the bandwidth to schedule
+and prioritize transactions is doubled.  There needs to be a design
+for BundleTransactions that allow the bundler to prioritize a batch
+of transactions for execution by paying a priority fee for all of
+them, and executing the whole batch together.
+
+Priority fees now also imply execution priority.
+
+Fork choice and voting is not blocked by user block propagation or
+execution of user transactions.
+
+Builders are going to capture all the MEV.
+
+Votes contain additional data, instead of BankHash, they contain
+(VoteHash, (BankHash, slot))
+
+## Security Considerations
+
+Validators may include incorrect BankHash values which may not
+be detected instantly.
+
+Validators may skip evaluating their own BankHash and copy results
+from other validators.
+
+Network halts if it cannot compute a epoch snapshot hash once an
+epoch because it is overloaded with user transactions.
+
+Leader could censor or delay the builder.
+
+Under heavy forking, validators will skip executing all the
+UserBlocks on minor forks.
+
+## Economic Considerations
+
+Builders should be shuffled and scheduled according to stake weight.
+
+TBD, deciding how should builders and leaders split the fees from
```

Comment:
Something to consider here is how to handle duplicate transactions.

Reply:
sorry might be dumb but don't we already have dedup features?

Reply:
No problem my previous message was not clear 😄 

We have deduping that prevents txs from appearing in multiple different slots. With this proposal, we have multiple builders who could potentially include the same transaction in their `UserBlock` since they can't know ahead of time what the other will include. We need to ignore one of them so we don't process the transaction twice, which is straight-forward to do.

My previous comment was mainly highlighting an edge-case for the economics of this deduplication. They both included a valid transaction in their `UserBlock`, so how do we handle fees?

Potentially a few options:
1) builder proportion of fees for tx split evenly between the two builders
2) first builder's block to reach leader

1 seems a better option, but it's also going to entirely depend how we split fees between the leader and builders in general.

Reply:
Could the transaction being sent contain information that specifying which leader to process it? That would avoid duplicates and it's easier to filter the block for transactions referencing one of the particular leaders to know which fee rewards it should receive.

Reply:
Maybe being able to specify the leader is also a feature though? Having your transaction processed by a certain leader might be undesirable for mev reasons?

Reply:
Wdym processed by a certain leader? Either the UserBlcok is included or it's not. The big question is if it can be included in any slot, or only the scheduled slot.

Reply:
I believe what is meant is a mechanism to make sure a transaction is processed by a specific **builder**

> Having your transaction processed by a certain leader might be undesirable for mev reasons so there could be reason for specificity?

If there's a certain builder I know does some MEV that is undesirable for my transaction, then I would request my tx is processed by the other.

Reply:
Sorry again a dumb idea
Is there any way to encrypt the txns beforehand and reveal the transaction details only after they are included in the block to prevent bad reordering? Homomorphic encryption could allow operations to be run on the encrypted builder blocks

Reply:
> I believe what is meant is a mechanism to make sure a transaction is processed by a specific **builder**
> 
> > Having your transaction processed by a certain leader might be undesirable for mev reasons so there could be reason for specificity?
> 
> If there's a certain builder I know does some MEV that is undesirable for my transaction, then I would request my tx is processed by the other.

if the builder selection was based on the first few bits of the signature:

* builders could easily filter duplicate transactions submitted to the wrong builder
* power users could choose builders by bit-flipping data in their message
* users who don't care get properly load shedded